feat(eslint-plugin): add prefer-regexp-exec rule (#305)
typescript-eslint/typescript-eslint: packages/eslint-plugin/docs/rules/prefer-regexp-exec.md

# Enforce to use `RegExp#exec` over `String#match` (prefer-regexp-exec)

> `String#match` is outperformed by `RegExp#exec` and both work in the same way, except for /g flag.

Reviewer: Would be useful to link to the source of that statement. (see https://github.com/typescript-eslint/typescript-eslint/issues/283)

Reviewer: ```suggestion
`RegExp#exec` is faster than `String#match` and both work the same when not using the `/g` flag.
```

## Rule Details

This rule is aimed at enforcing the more performant way for applying regular expressions on Strings.

Examples of **incorrect** code for this rule:

```ts
'something'.match(/thing/);

const text = 'something';

> const search = /thing/;

Reviewer: It's inconsistent that you use variables here, but not in the "correct" example.

text.match(search);
```

Examples of **correct** code for this rule:

```ts
'some things are just things'.match(/thing/g);
/thing/.exec('something');
```

## Options

There are no options.

```JSON
{
  "@typescript-eslint/prefer-regexp-exec": "error"
}
```

## When Not To Use It

> If you don't mind performance, you can turn this rule off safely.

Reviewer: This could be written better.
